staging/src/kubesphere.io/utils/helm/executor.go

```diff
@@ -36,41 +56,488 @@ type Executor interface {
-	return "", nil
+	sts, err := e.status()
+	if err != nil {
+		return "", err
+	}
+
+	if sts.Info.Status == "deployed" {
+		return e.createInstallJob(ctx, chartName, chartData, values, true)
+	}
+	return "", fmt.Errorf("cannot upgrade release %s/%s, current state is %s", e.Namespace, e.ReleaseName, sts.Info.Status)
+}
+
+func (e *executor) kubeConfigPath() string {
+	if len(e.KubeConfig) == 0 {
+		return ""
+	}
+	return "kube.config"
+}
+
+func (e *executor) chartPath(chartName string) string {
+	return fmt.Sprintf("%s.tgz", chartName)
+}
+
+func (e *executor) setupChartData(chartName, chartData, values string) (map[string]string, error) {
+	if len(e.labels) == 0 && len(e.annotations) == 0 {
+		return nil, nil
+	}
+
+	kustomizationConfig := types.Kustomization{
+		Resources:         []string{"./.local-helm-output.yaml"},
+		CommonAnnotations: e.annotations,                    // add extra annotations to output
+		Labels:            []types.Label{{Pairs: e.labels}}, // Labels to add to all objects but not selectors.
+	}
+	kustomizationData, err := yaml.Marshal(kustomizationConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	data := map[string]string{
+		postRenderExecFile:     kustomizeBuild,
+		kustomizationFile:      string(kustomizationData),
+		e.chartPath(chartName): chartData,
+		"values.yaml":          values,
+	}
+	if e.kubeConfigPath() != "" {
+		data[e.kubeConfigPath()] = e.KubeConfig
+	}
+	return data, nil
+}
+
+func (e *executor) createConfigMap(ctx context.Context, chartName, chartData, values string) (string, error) {
+	data, err := e.setupChartData(chartName, chartData, values)
+	if err != nil {
+		return "", err
+	}
+
+	name := fmt.Sprintf("helm-executor-%s-%s", chartName, rand.String(6))
+	configMap := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: e.Namespace,
+		},
+		Data: data,
+	}
+	if err = e.Create(ctx, configMap); err != nil {
+		return "", err
+	}
+	return name, nil
+}
+
+func (e *executor) createInstallJob(ctx context.Context, chartName, chartData, values string, upgrade bool) (string, error) {
+	args := make([]string, 0, 10)
+	if upgrade {
+		args = append(args, "upgrade")
+	} else {
+		args = append(args, "install")
+	}
+
+	args = append(args, "--wait", e.ReleaseName, e.chartPath(chartName), "--namespace", e.Namespace)
+
+	if len(values) > 0 {
+		args = append(args, "--values", "values.yaml")
+	}
+
+	if e.dryRun {
+		args = append(args, "--dry-run")
+	}
+
+	if e.kubeConfigPath() != "" {
+		args = append(args, "--kubeconfig", e.kubeConfigPath())
+	}
+
+	// Post render, add annotations or labels to resources
+	if len(e.labels) > 0 || len(e.annotations) > 0 {
+		args = append(args, "--post-renderer", filepath.Join(workspaceBase, postRenderExecFile))
+	}
+
+	if klog.V(8) {
+		// output debug info
+		args = append(args, "--debug")
+	}
+
+	name, err := e.createConfigMap(ctx, chartName, chartData, values)
+	if err != nil {
+		return "", err
+	}
+	job := &batchv1.Job{
```

Comment:
It's better to use an embed template file instead.

Reply:
I don't think so, first of all, the content of this template cannot be reused, the installation/upgrade and uninstallation need different things, and secondly, the content of the `Job` may need to be dynamically added or subtracted according to the parameters, using a template does not make this easier to read or maintain.